### PR TITLE
feat(gatsby): Support incremental builds with collection routes

### PR DESCRIPTION
**Pull request by blainekasten**
## Description

This PR adds support for Gatsby incremental builds with collection routes. Included in here was quite a large file refactor. If this PR is too big I can do that in a different PR. At a high level what this PR does is:

- Adds an `onDeleteNode` lifecycle method to gatsby-node 
- Move the collection route building _out of_ `createPagesStatefully` and into `createPages`
- Add `onDeleteNode` to delete pages from a node that is removed if a page exists for it

## Related Issues

 #26375

### CODE REVIEW

**Comment by wardpeet** [comment]
Blocked it for now cause onDeleteNode needs discussion

File: packages/gatsby-telemetry/src/telemetry.ts [hunk 1]
```diff
@@ -426,6 +426,7 @@ export class AnalyticsTracker {
   }
 
   trackFeatureIsUsed(name: string): void {
+    if (this.features.has(name)) return
```
[the remark this answers is not in the document]
Comment (pieh) [reply packages/gatsby-telemetry/src/telemetry.ts:429]: Can we split this fix from this PR and have it as standalone PR? (this PR adding new API lifecycle will likely take some time)

**Comment by blainekasten** [comment]
PR is no longer blocked as consensus has been given to implement the `onDeleteNode` API

**Comment by blainekasten** [comment]
Turns out we do not need this code. Incremental Builds already magically works with collection routes!

**Pull request closed without merging**